query-engine/core/src/query_graph_builder/extractors/utils.rs

```diff
@@ -1,24 +1,23 @@
 use crate::schema_builder;
-use prisma_models::{ModelRef, ScalarFieldRef};
+use prisma_models::{IndexField, ModelRef};
 
 /// Attempts to resolve a field name to a compound field.
-pub fn resolve_compound_field(name: &str, model: &ModelRef) -> Option<Vec<(Vec<String>, ScalarFieldRef)>> {
+pub fn resolve_compound_field(name: &str, model: &ModelRef) -> Option<Vec<IndexField>> {
     resolve_compound_id(name, model).or_else(|| resolve_index_fields(name, model))
 }
 
 /// Attempts to match a given name to the (schema) name of a compound id field on the model.
-pub fn resolve_compound_id(name: &str, model: &ModelRef) -> Option<Vec<(Vec<String>, ScalarFieldRef)>> {
+fn resolve_compound_id(name: &str, model: &ModelRef) -> Option<Vec<IndexField>> {
     model.fields().compound_id().and_then(|pk| {
-        (name == schema_builder::compound_id_field_name(pk))
-            .then(|| pk.fields().into_iter().map(|f| (vec![], f)).collect())
+        (name == schema_builder::compound_id_field_name(pk)).then(|| IndexField::from_scalars(pk.fields()))
     })
 }
 
 /// Attempts to match a given name to the (schema) name of a compound indexes on the model and returns the first match.
-pub fn resolve_index_fields(name: &str, model: &ModelRef) -> Option<Vec<(Vec<String>, ScalarFieldRef)>> {
+fn resolve_index_fields(name: &str, model: &ModelRef) -> Option<Vec<IndexField>> {
     model
         .unique_indexes()
         .into_iter()
         .find(|index| schema_builder::compound_index_field_name(index) == name)
-        .map(|index| index.fields())
+        .map(|index| index.fields().to_vec())
 }
```

Comment:
Fought the compiler to not clone the indexes on line 22 but doesn't seem like I can do anything.
Since I have to instantiate new index fields in `resolve_compound_id` (see line 12`), I can't make that function (`resolve_compound_id`) return a reference on those newly instantiated index fields.
Therefore, those three functions have to return `Vec<IndexField>` instead of `&[IndexField]`.
If you see a way to make that work, please let me know 🙏